### DISCUSSION TITLE
Use of hard tabs in code style ?

### DISCUSSION THREAD

**Discussion by breadoven** [iNavFlight/inav#9837] [General]
I've noticed there seems to be a lot of hard tabs used in PR's that have been merged recently. This screws up the formatting of the code when viewed on Github making it hard to read. I thought hard tabs were not to be used according to:

https://github.com/multiwii/baseflight/wiki/CodingStyle

Has this requirement been relaxed ?

**Comment by MrD-RC** [comment]
Not as far as I know. Tabs should be used for indentation, not spaces.

**Comment by breadoven** [comment]
You can use Tabs, just not hard tabs, soft Tabs should be used, i.e. padded with spaces. Your code editor should be set up to make a Tab 4 spaces, a soft tab. If you don't do this then the file viewer will tabulate the code according to its tabulation settings which may not correspond to 4 spaces. Using a soft Tab ensures consistency because a space should be consistent across viewers.

The affect of hard tabs can be seen here with the indentation a mess:
![Screenshot (155)](https://github.com/iNavFlight/inav/assets/56191411/b8e42dd5-f8f6-4655-83c5-17fd9b3a308c)
The arrows are hard Tabs:
![Screenshot (156)](https://github.com/iNavFlight/inav/assets/56191411/de2aa8c7-a942-4293-9d25-f47e07f983aa)

It should look like this to avoid this problem, i.e. just spaces:
![Screenshot (158)](https://github.com/iNavFlight/inav/assets/56191411/b3041331-84e0-4b35-99c2-fb9d92280815)

I seem to remember the Workflow checker used to fail if hard tabs were used. Obviously this doesn't happen any more.

**Comment by MrD-RC** [comment]
Damn, just realised I'm a perp here! VSCode uses hard tabs by default. I've just changed the setting. In the settings, search for `Insert spaces`. It will return *Editor: **Insert Spaces*** Turn that off, and it stops with the hard spaces.

**Comment by stronnag** [comment]
Unfortunately its just too easy to  go with `$EDITOR`'s default C-style. Mea culpa, hard tab fan and several other bracket related sins contrary to the style guide. I used to have a `.emacs` rule to put me on the straight and narrow, but ... just another offender.

**Comment by mmosca** [comment]
> Unfortunately its just too easy to go with `$EDITOR`'s default C-style. Mea culpa, hard tab fan and several other bracket related sins contrary to the style guide. I used to have a `.emacs` rule to put me on the straight and narrow, but ... just another offender.

I am not against having .emacs .vimrc files in the repo to try to enforce desired behavior as default.

**Comment by MrD-RC** [comment]
A nice little tip for VSCode

To convert existing indentation from spaces to tabs hit Ctrl+Shift+P and type: `Convert indentation to Tabs`

**Comment by RomanLut** [comment]
Note: Please do not waste time fixing tabs in HITL-related code; they are fixed in PR: https://github.com/iNavFlight/inav/pull/8347
(sorry noticed tab problem too late)
